src/sentry/db/models/outboxes.py

```diff
@@ -0,0 +1,141 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, Collection, List, Mapping, Type
+
+from sentry.types.region import find_regions_for_orgs
+
+if TYPE_CHECKING:
+    from sentry.models.outbox import ControlOutboxBase, OutboxCategory, RegionOutboxBase
+
+
+class ProducesRegionOutboxOnUpdate:
+    __default_flush__ = True
+
+    def outbox_for_update(self, shard_identifier: int | None = None) -> RegionOutboxBase:
+        raise NotImplementedError
+
+
+class ProcessUpdatesWithRegionOutboxes(ProducesRegionOutboxOnUpdate):
+    __category__: OutboxCategory
+    __outbox_type__: Type[RegionOutboxBase] | None = None
```

Comment:
Nit, but I'd like to check our convention on the double-underscore affixes. Should "double underscore before and after" be reserved for language-defined methods such as `__eq__`?

Personally, I've been using "double underscore before but *not* after" in places like `rpc.py` to mean, "Special attribute on a base class that's reserved from subclasses and is going to control behavior elsewhere." (In contrast to a single-underscore prefix, which means code outside the class should not look at it at all.)

This seems like a comparable case. But like I said, that's just my personal semantics, not necessarily a general convention.